src/Symfony/Component/Process/InputStream.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Process;
+
+use Symfony\Component\Process\Exception\RuntimeException;
+
+/**
+ * Provides a way to continuously write to the input of a Process until the InputStream is closed.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class InputStream implements \IteratorAggregate
+{
+    private $onEmpty = null;
+    private $input = array();
+    private $open = true;
+
+    /**
+     * Sets a callback that is called when the write buffer is emptied.
+     */
+    public function onEmpty(callable $onEmpty = null)
+    {
+        $this->onEmpty = $onEmpty;
+    }
+
+    /**
+     * Appends an input to the write buffer.
+     */
+    public function write($input)
+    {
+        if (null === $input) {
+            return;
+        }
+        if ($this->isClosed()) {
+            throw new RuntimeException(sprintf('%s is closed', get_class($this)));
+        }
+        $this->input[] = ProcessUtils::validateInput(__METHOD__, $input);
+    }
+
+    /**
+     * Closes the write buffer.
+     */
+    public function close()
+    {
+        $this->open = false;
+    }
+
+    /**
+     * Tells is write buffer is closed or not.
+     */
+    public function isClosed()
+    {
+        return !$this->open;
+    }
+
+    public function getIterator()
+    {
+        $this->open = true;
+
+        while ($this->open || $this->input) {
+            if (!$this->input) {
```

Comment:
The yield is what breaks the infinite loop: the state can change between two iterations. But you know that.
About your suggestion, that's how it was in an earlier version, but I ended up preferring this semantic for onEmpty: to be called when the buffer *becomes* empty.
Otherwise, the callback would be called both for initializing the buffer (assuming it started empty) and repeatedly when it's empty again, which makes it very hard to reliably know how many times it will be called.
In the submitted version, the reasonability looks more clear to me: repopulated the buffer when it's getting empty.
That's how I thought about it at least.